src/main/host/network/queuing.rs

```diff
@@ -0,0 +1,321 @@
+    AlreadyQueued,
+    /// A priority value was not provided when one is required.
+    NonePriority,
+}
+
+/// A queue that can hold a group of items (e.g., sockets or packets) and that provides some control
+/// over the queuing discipline and behavior. The queue stores two copies of every enqueued item:
+/// the items and their clones must be such that they are considered equal. Thus, it may be useful
+/// to wrap items with a `std::rc::Rc` or `std::sync::Arc` before enqueuing.
+pub struct NetworkQueue<T: Clone + Debug + Eq + Hash> {
+    /// The set of items that currently exist in the queue.
+    membership: HashSet<T>,
+    /// A cloned set of items organized according to the configured queuing discipline.
+    queue: QueuingDiscipline<T>,
+}
+
+impl<T: Clone + Debug + Eq + Hash> NetworkQueue<T> {
+    /// Create a new queue that will organize items according to the given `NetworkQueueKind`.
+    pub fn new(kind: NetworkQueueKind) -> Self {
+        let queue = match kind {
+            NetworkQueueKind::MinPriority => QueuingDiscipline::MinPriority(BinaryHeap::new()),
+            NetworkQueueKind::FirstInFirstOut => {
+                QueuingDiscipline::FirstInFirstOut(VecDeque::new())
+            }
+        };
+        Self {
+            membership: HashSet::new(),
+            queue,
+        }
+    }
+
+    /// Remove and drop all items from the queue.
+    pub fn clear(&mut self) {
+        match &mut self.queue {
+            QueuingDiscipline::MinPriority(heap) => heap.clear(),
+            QueuingDiscipline::FirstInFirstOut(deque) => deque.clear(),
+        }
+        self.membership.clear();
+    }
+
+    /// Returns true if the queue contains the item, and false otherwise.
+    pub fn contains(&self, item: &T) -> bool {
+        self.membership.contains(item)
+    }
+
+    /// Returns true if the queue does not contain any items.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the number of queued items.
+    pub fn len(&self) -> usize {
+        self.membership.len()
+    }
+
+    /// Returns the next available item if one exists. This function also drops the duplicate cloned
+    /// handle that was created in `NetworkQueue::try_push()`.
+    pub fn pop(&mut self) -> Option<T> {
+        match &mut self.queue {
+            QueuingDiscipline::MinPriority(heap) => heap.pop().map(|x| x.item),
+            QueuingDiscipline::FirstInFirstOut(deque) => deque.pop_front(),
+        }
+        .inspect(|x| {
+            assert!(self.membership.remove(x));
+        })
+    }
+
+    /// Pushes an item into the queue, internally cloning it to
+    /// support membership checks. It may be useful to wrap items with a `std::rc::Rc` or
+    /// `std::sync::Arc` before calling this function.
+    ///
+    /// # Panics
+    /// This function panics if the queue was configured with `NetworkQueueKind::MinPriority` and
+    /// `maybe_priority` is `None`.
```

Comment:
I think this might need an extra newline, or else markdown might say that this is all part of the heading?

Reply:
I checked a couple of renderers and it seems OK as-is:
- https://markdownlivepreview.com
- https://www.digitalocean.com/community/markdown

It also renders correctly in VSCode, and on Github.

That being said, I guess it's better to be safe and just add the extra newline :)